### DISCUSSION TITLE
Help needed, clarification on PrimaryDataSource Option used by SQL4Cds Engine.

### DISCUSSION THREAD

**Discussion by ainsleybernard** [MarkMpn/Sql4Cds#149] [Q&A] [answered]
Hello,

What should we be putting for primaryDataSource for the options, I copied the value from test project which was set to "local". I didnt think that was right but im not sure what value it expects.

But I get the following error, when I running the following sql "select top 1 name from account "
![image](https://user-images.githubusercontent.com/17305314/148583103-3c7edcdd-fb3b-44b1-ae22-0c4886d922fb.png)

**Comment by MarkMpn** [comment, marked as answer]
In the `ExecutionPlanBuilder` constructor you pass in a list of data sources, each of which has a name. The `PrimaryDataSource` should be the name of one of these data sources.

When you refer to a table as just `account`, it will look for that table within the primary data source. If you refer to it as `datasource.dbo.account`, it will look for it within that named data source instead.

**Reply by ainsleybernard** [reply]
Thanks for reply @MarkMpn!

Setting the `primaryDataSource` worked for me.

But the other example is a little unclear to me.


If I have following

```

    var dataSource = new DataSource
            {
                Name = "prod",
                Connection = serviceClient,
                Metadata = metadata,
                TableSizeCache = new TableSizeCache(serviceClient, metadata)
            };
```

my sql could be?

`            var sql = "select top 1 accountid from prod.account ";
`

**Reply by MarkMpn** [reply]
Not quite - it would be:

```sql
select top 1 accountid from prod.dbo.account
```

Standard data tables are in the `dbo` schema. You can also access metadata in the `metadata` schema, e.g.

```sql
select logicalname from prod.metadata.entity
```

**Reply by ainsleybernard** [reply]
ohhh I gotcha, thanks!!! Very cool stuff :)
